ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py

```diff
@@ -0,0 +1,244 @@
+    )
+    logger.info("loaded and filtered crunchbase data")
+
+    gtr = load_s3_data(bucket_name, "outputs/gtr/gtr_ai_genomics_projects.csv")
+    gtr_filtered = filter_data(
+        data=gtr, query="ai_genomics == True", date_col="start", id_col="id"
+    )
+    logger.info("loaded and filtered gtr data")
+
+    oa = get_openalex_ai_genomics_works()
+    oa_filtered = filter_data(
+        data=oa,
+        query="ai_genomics == True",
+        date_col="publication_date",
+        id_col="work_id",
+    )
+    logger.info("loaded and filtered oa data")
+
+    # timeslice entities
+    ents_per_date = timestamp_entities(
+        list_of_dfs=[patents_filtered, crunchbase_filtered, gtr_filtered],
+        list_of_ents=[patent_ents, crunchbase_ents, gtr_ents],
+    )
+    logger.info("timesliced entities on a yearly basis from 2015 onwards.")
+
+    # embed and reduce ents and generate lookup
+    all_ents = list(set(list(itertools.chain(*list(ents_per_date.values())))))
+    ent_embeds_lookup = generate_entities_embedding_lookup(entities=all_ents)
+    logger.info("generated reduced entity embedding lookup.")
+
+    # identify optimal k at every timeslice
+    best_ks = get_best_k(
+        reduced_entities=ent_embeds_lookup, ents_per_date=ents_per_date
+    )
+    logger.info("got best k per timesliced entities.")
+
+    # cluster entities with optimal k at each timeslice
+    ents_per_date_clusts = dict()
+    for best_k, (year, ents) in zip(best_ks, ents_per_date.items()):
+        ent_embeds = [ent_embeds_lookup.get(ent) for ent in ents]
+        km = KMeans(n_clusters=best_k).fit(ent_embeds)
+        clust = km.predict(ent_embeds)
+        clust_dict = {c: [] for c in clust}
+        for i, c in enumerate(clust):
+            clust_dict[c].append(ents[i])
+        ents_per_date_clusts[year] = clust_dict
+    logger.info("clustered entities at each timestamp using best ks.")
+
+    #### propogate cluster names across timeslices
+    years = list(ents_per_date_clusts.keys())
+    for year in range(len(years) - 1):
+        timeslice_x = ents_per_date_clusts[years[year]]
+        timeslice_y = ents_per_date_clusts[years[year + 1]]
+        propagate_labels(timeslice_x, timeslice_y)
+
+    logger.info("propagated labels across timeslices.")
+
+    # save look up in json serializable way 
+    save_to_s3(
+        bucket_name,
+        {k:v.tolist() for k, v in ent_embeds_lookup.items()},
+        "outputs/analysis/tag_evolution/dbpedia_reduced_embeds.json",
+    )
+    # change keys to string to be able to save dict
+    save_to_s3(
+        bucket_name,
+        {
+            year: {str(k): v for k, v in ent_info.items()}
+            for year, ent_info in ents_per_date_clusts.items()
+        },
+        "outputs/analysis/tag_evolution/dbpedia_clusters_timeslice_embed.json",
+    )
+    logger.info("saved dbpedia entities and propagated entity clusters.")
```

Comment:
Where are you saving the timeslices with propagated names? I thought those were stored in `timeslice_x` and `timeline_y`? Or are the changes to that propagating to `ents_per_date_clusts` somehow? I would try to make that a bit clearer.

Reply:
the changes happen in place to ents_per_date_clusts - I've added comments so its easier to follow